### PR TITLE
Improve search param extraction flexibility

### PR DESCRIPTION
**Pull request by jmarchionatto**
What was done: 

1. added `myExtractResourceLevelParams` property to allow extraction of Resource-level search param values 
2. extracted some code to be reused to an util class
3. Allowed test to specify a Patient provider to `RestServerR4Helper`

### CODE REVIEW

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java [hunk 1]
```diff
@@ -2535,4 +2530,8 @@ public void extract(
 			myExtractor1.extract(theParams, theSearchParam, theValue, thePath, theWantLocalReferences);
 		}
 	}
+
+	public void setExtractResourceLevelParams(boolean theExtractResourceLevelParams) {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:2534]: I don't see any tests that exercise this behaviour. Should probably have a test with positive and negative values.

Comment (jmarchionatto) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractor.java:2536]: Added

**Review by dotasek — CHANGES_REQUESTED** [review]
Test for the myExtractResourceLevelParams looks good. Is this change attached to a feature, or something else that needs to be put in the changelog, or are these just quality of life improvements.

File: hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractorTest.java [hunk 2]
```diff
@@ -1,18 +1,91 @@
...
+
+			// execute
+			myExtractor.extractSearchParams(myResource, null, mySearchParamType, false, mySearchParamFilter);
+
+			verify(myExtractor).extractSearchParam(any(), any(), any(), any(), eq(false));
+		}
+
+		@Test
+		void testWhenNotSet_resourceLevelParamsAreNotExtracted() {
+			doReturn(List.of(myRuntimeSearchParam)).when(myExtractor).getSearchParams(myResource);
```
[the remark this answers is not in the document]
Comment (jmarchionatto) [reply hapi-fhir-jpaserver-searchparam/src/test/java/ca/uhn/fhir/jpa/searchparam/extractor/BaseSearchParamExtractorTest.java]: Done. Thanks

**Review by jamesagnew — APPROVED** [review]

**Review by dotasek — APPROVED** [review]

**Pull request merged**